### PR TITLE
Ensured index cache is best effort, refactored tests, validated edge cases.

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes https://github.com/improbable-eng/thanos/issues/651

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

cc @mjd95

### CODE REVIEW

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -225,10 +226,18 @@ func NewBucketStore(
 		return nil, errors.Errorf("max concurrency value cannot be lower than 0 (got %v)", maxConcurrent)
 	}
 
-	indexCache, err := newIndexCache(reg, indexCacheSizeBytes)
+	// TODO(bwplotka): Add as a flag?
+	maxItemSizeBytes := uint64(float64(indexCacheSizeBytes) * 0.5)
```
Comment (bwplotka) [inline pkg/store/bucket.go]: Assumption: Max item size should be not bigget than 50% of Overall size. I think it's generous already.

File: pkg/store/bucket.go [hunk 2]
```diff
@@ -225,10 +226,18 @@ func NewBucketStore(
...
+
+	indexCache, err := storecache.NewIndexCache(logger, reg, storecache.Opts{
+		SetTimeout:       300 * time.Millisecond, // TODO(bwplotka): Add as a flag?
```
Comment (bwplotka) [inline pkg/store/bucket.go]: arbitrary number, should be enough for typical setup?

File: pkg/store/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+		reg.MustRegister(prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+			Name: "thanos_store_index_cache_max_item_size_bytes",
+			Help: "Maximum number of bytes for single entry to be held in the index cache.",
+		}, func() float64 {
+			return float64(c.maxItemSizeBytes)
+		}))
+		reg.MustRegister(c.requests, c.hits, c.added, c.evicted, c.current, c.currentSize, c.overflow, c.cancelled)
+	}
+
+	if opts.MaxSizeBytes > 1e12 {
```
Comment (devnev) [inline pkg/store/cache/cache.go]: Isn't the `1e12` the maximum number of _items_ in the LRU, unrelated to their size?

Comment (bwplotka) [reply pkg/store/cache/cache.go]: You are right, changing the comment (and removing this check)

File: pkg/store/bucket.go (hunk 1 above)
Comment (devnev) [inline pkg/store/bucket.go]: `maxItemsSizeBytes = indexCacheSizeBytes / 2`

Comment (bwplotka) [reply pkg/store/bucket.go]: lol

**Comment by bwplotka** [comment]
PTAL @devnev

File: pkg/store/cache/cache.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+		return cacheTypeSeries
+	}
+	return "<unknown>"
+}
+
+func (c cacheItem) size() uint64 {
+	switch k := c.key.(type) {
+	case cacheKeyPostings:
+		// ULID + 2 slice headers + number of chars in value and name.
+		return sliceHeaderSize + 2*16 + uint64(len(k.Value)+len(k.Name))
```
Comment (devnev) [inline pkg/store/cache/cache.go]: Comment and code don't quite match up ("ulid + 2 slice headers" vs `sliceHeaderSize + 2*16`).

File: pkg/store/cache/cache.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+type cacheKeySeries uint64
+
+type IndexCache struct {
+	mtx sync.Mutex
+
+	logger           log.Logger
+	lru              *lru.LRU
+	maxSizeBytes     uint64
+	maxItemSizeBytes uint64
+	setTimeout       time.Duration
```
Comment (devnev) [inline pkg/store/cache/cache.go]: `setTimeout` unused, remove

File: pkg/store/cache/cache.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+		"maxSizeBytes", c.maxSizeBytes,
+		"setTimeout", c.setTimeout.String(),
+		"maxItems", "math.MaxInt64",
+	)
+	return c, nil
+}
+
+func (c *IndexCache) onEvict(key, val interface{}) {
+	k := key.(cacheItem).keyType()
+	entrySize := sliceHeaderSize + uint64(len(val.([]byte)))
```
Comment (devnev) [inline pkg/store/cache/cache.go]: Duplicating the entrySize calculation everywhere seems dubious. It's already inconsistent, sometimes using a literal `16` and sometimes `sliceHeaderSize`.

File: pkg/store/cache/cache.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+)
+
+const (
+	cacheTypePostings string = "Postings"
+	cacheTypeSeries   string = "Series"
+
+	sliceHeaderSize = 16
+)
+
+type cacheItem struct {
```
Comment (devnev) [inline pkg/store/cache/cache.go]: Reading the code is made somewhat more confusing by this being called `cacheItem` when it is used as the cache key. WDYT about renaming to `cacheKey`?

File: pkg/store/cache/cache.go (hunk 6 above)
Comment (bwplotka) [reply pkg/store/cache/cache.go]: sorry, might miss some places ):

File: pkg/store/cache/cache.go (hunk 7 above)
Comment (bwplotka) [reply pkg/store/cache/cache.go]: yes!

File: pkg/store/cache/cache.go (hunk 6 above)
Comment (bwplotka) [reply pkg/store/cache/cache.go]: what do you mean duplicating though?

Comment (devnev) [reply pkg/store/cache/cache.go]: There are three locations where you calculate `entrySize`, all of which must do the same calculation for the cache to work correctly. It might be worth factoring this out, although I guess it might be good enough as is.
